### ISSUE TITLE
`pipenv install <package>` with a CLI flag to install the most recent version and have it save to the Pipfile

### ISSUE THREAD

**Issue by abarbet** [pypa/pipenv#5531]
### Is your feature request related to a problem? Please describe.

When running `pipenv install <package>` without specifying a version, the Pipfile stores the reference to the package as `<package> = "*"`. While this is perfectly functional, anytime a new lockfile is generated, the most recent version of that package will be pulled into the new dependency tree. A version update like this could potentially be a breaking change for the project.

Of course, I could manually go to PyPI at the time of install and manually specify the version, but if I know it works at that version number, it can be a lot of extra work to find that most recent version number and specify it in the command line (especially if I'm bulk installing a lot of packages at once).

### Describe the solution you'd like

Something like an additional flag when running `pipenv install` would be ideal, i.e., `pipenv install --lock-recent <package>` would fetch and install the most recent package, but ALSO save the explicit reference to the most recent version in the Pipfile. 

### Describe alternatives you've considered

Alternative would be the manual lookup in PyPI.

-------------------------------------------------------------------------------

**Comment by lroberts7** [comment]
stack overflow question with the same request:

https://stackoverflow.com/questions/70725867/pipenv-install-package-with-exact-latest-version

indicates this feature request would be used by a large portion of the `pipenv` users.

Is this a contribution that the `pipenv` maintainers would welcome as a CR?

**Comment by mmisiewicz** [comment]
I would like this feature as well!

**Comment by oz123** [comment]
I think it's a reasonable behavior and I welcome a PR for always pinning the latest version.
However,  it is a behavioral change so this needs to be added with a new flag as suggested.

**Comment by ryanovas** [comment]
I'm a little surprised this isn't the default behaviour.... This way I might just run "pipenv install" and accidentally install a new major version. It's super annoying to have to pip freeze every time and manually set the version.

**Comment by oz123** [comment]
I agree.. It's a weird behavior of pipenv.. unfortunately it's something we inherited. As I said, PR welcome.

**Comment by tadasajon** [comment]
I'm honestly flabbergasted that this isn't the default behavior

**Comment by jatinb92** [comment]
Hi, @matteius @oz123 I am trying to work on it, but I have a doubt and need your confirmation before proceeding ahead.

If we want to use this flag "--lock-recent",  which means we should only install the latest version of the new package but all the rest of the packages should not get updated and only install the version mentioned in the pipfile. 

So, when we are updating the packages using the "get_version" method in utils.py, we should receive an exact number and not ''(blank). For that, we have to save that version in the pipfile and not '*'. 

Am I correct in the above assumption and should continue on this path?

**Comment by matteius** [comment]
I am not convinced that the answer is a flag named `--lock-recent`, I would take issue with the name if we do go with an install flag.

Let's start with high level -- is it better for this to be a flag issued with the install command, or is it better to be a project wide option that gets put in the pipenv section of the Pipfile?   Some people are saying they can't believe its not the default, I can believe it because the default not to pin makes sense given you can specify the version you want and it pins to that, so what we are discussing is really a convenience measure to not have to look up the latest dependencies version, and to pin to the latest version.

If we did go with an install flag, I'd want us to think more about the best name.   Some thoughts (maybe not best) `--pin-version` might be better -- in both cases the lock file gets pinned so calling it `--lock` anything would be a misnomer to me.   If its a flag, it should not be overly verbose either so its easy to type.   Is `--pin` too generic?

**Comment by matteius** [comment]
I've been working on this tool for a year and a half now and it has never been my intention to emulate npm--I want to throw out that misconception that pipenv is somehow analogous to npm, it has some parallels sure, but it is its own thing.

Most devs I've worked with using pipenv actually don't want to pin to a very specific version, they want to pin to a major version, and so this idea of changing the default behavior to pin to the latest version of something definitely will not make everyone happy.   Also changing default behaviors in general should be carefully planned out; whenever we have changed what people expect there has been varying degrees of push back whether the change to the default behavior was even intended or not.  That being said, I've become more comfortable making behavioral changes to pipenv and I ask that we think through all of the use cases.   

I generally pin a lot of things to * in my personal projects and even pipenv itself pips most of its dependencies to "*" where possible so that we are always pushing for the latest upgrade paths, only things that have caused issues and needed to be pinned to a prior version for us are.   By making the default behavior to pin to what you installed initially, you create more work for yourself when re-locking, assuming you want to be on the latest versions of everything to get the bugfixes and latest features.

I don't think that its dangerous or unexpected, as this has always been the default behavior, so I don't agree with the premise that keeping * as the default doesn't make sense.

I still want to work with the community to arrive at the best flags and interfaces for all the various use cases, and from your write up to me that sounds like a vote for a project wide [pipenv] directive that would make pinning to the current latest the default behavior for your project users.  This would be more useful than expecting everyone to pass the same flag every time.

**Comment by lroberts7** [comment]
> Let's start with high level -- is it better for this to be a flag issued with the install command, or is it better to be a project wide option that gets put in the pipenv section of the Pipfile? Some people are saying they can't believe its not the default, I can believe it because the default not to pin makes sense given you can specify the version you want and it pins to that, so what we are discussing is really a convenience measure to not have to look up the latest dependencies version, and to pin to the latest version.

I agree that breaking backward compatibility here is unnecessary. 

I can believe it's the default b/c that's the current behavior for me locally :) 

Also agree this is really a convenience measure request. 

> is it better to be a project wide option that gets put in the pipenv section of the Pipfile?

I hadn't considered this approach and I agree with the suggestion, adding it into the pipenv section of the Pipfile is preferred. 





also, @matteius thanks for your work on `pipenv`-it's a great package.

**Comment by matteius** [comment]
I am imagining there may be a way to specify the [pipenv] directive to allow users to customize if they want specific, major, or major + minor pins.   If someone wants to take this on, feel free to take a shot and open a PR.  I may not be able to get to it for a bit.

**Comment by jatinb92** [comment]
> is it better to be a project wide option that gets put in the pipenv section of the Pipfile?

I agree, using it in the file is much better to approach than using a flag. 

> So, when we are updating the packages using the "get_version" method in utils.py, we should receive an exact number and not ''(blank). For that, we have to save that version in the pipfile and not '*'.

But I am still looking at how to get which was the latest version they pinned. We need to save it somewhere in lockfile or pipfile, (lockfile I would prefer)
I am currently trying to work at a very basic change, post which we can improve the functionality.

Again it is my first time working on this project, so I would need your guidance for moving ahead @matteius .

**Comment by oz123** [comment]
I was all in favor of pinning the latest version as the default. But after reading the arguments by @matteius I changed my mind.
I feel now that a Project setting (saved in Pipfile) is the right way to go.

**Comment by matteius** [comment]
@jatinb92 the lock file already saves very specific version of whatever is locked, the goal here is a new directive in the `Pipfile` that would allow pinning to either the latest full version, latest major, or latest major + minor.    I would not recommend  working on this change as an entry point to contributing to pipenv as I expect it to be non-trivial, but if you figure it out feel free to open a PR.

**Comment by lemuelkn** [comment]
@jatinb92 I would like to tag along and help out with this problem, if you don't mind

**Comment by kalebmckale** [comment]
Perhaps I'm misunderstanding the issue and how it differs from employing `pipenv sync` or `--ignore-pipfile` as documented [here](https://pipenv-fork.readthedocs.io/en/latest/advanced.html):

> Or you can install packages exactly as specified in Pipfile.lock using the sync command:
> `$ pipenv sync`
> Note
> pipenv install --ignore-pipfile is nearly equivalent to pipenv sync, but pipenv sync will never attempt to re-lock your dependencies as it is considered an atomic operation. pipenv install by default does attempt to re-lock unless using the --deploy flag.

**Comment by kalebmckale** [comment]
I think I understand now after reading issue #5575. I forgot that the `Pipfile.lock` has multiple versions and hashes for various dependencies. In this case, this seems to be a potentially related issue or rather I have some rough ideas how there might be a solution that handles both.

**Comment by oz123** [comment]
@kalebmckale Pipfile and Pipfile.lock serve different purposes:
```
Pipfile -> for people
Pipfile.lock -> for pipenv itself to read and use. 
```

When you run `pipenv install`, it installs the packages and if a package isn't pinned with a specific version the latest version will be downloaded, installed and documented in `Pipfile.lock` only.

So, the idea of this issue is to never add packages with `foo="*"` to `Pipfile`.
